cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAutoRebalancingReconciler.java

```diff
@@ -0,0 +1,522 @@
+    private final List<KafkaAutoRebalanceConfiguration> kafkaAutoRebalanceConfigurations;
+    private final CrdOperator<KubernetesClient, KafkaRebalance, KafkaRebalanceList> kafkaRebalanceOperator;
+    private final Set<Integer> scalingDownBlockedNodes;
+
+    /**
+     * Constructs the Kafka auto-rebalancing reconciler
+     *
+     * @param reconciliation    Reconciliation marker
+     * @param kafkaCr   The Kafka custom resource
+     * @param supplier  Supplies the operators for different resources
+     * @param scalingDownBlockedNodes  nodes blocked on scaling down because of the need for auto-rebalancing first
+     */
+    public KafkaAutoRebalancingReconciler(
+            Reconciliation reconciliation,
+            Kafka kafkaCr,
+            ResourceOperatorSupplier supplier,
+            Set<Integer> scalingDownBlockedNodes) {
+        this.reconciliation = reconciliation;
+        this.kafkaAutoRebalanceStatus = kafkaCr.getStatus() != null ? kafkaCr.getStatus().getAutoRebalance() : null;
+        this.kafkaAutoRebalanceConfigurations = kafkaCr.getSpec().getCruiseControl().getAutoRebalance();
+        this.kafkaRebalanceOperator = supplier.kafkaRebalanceOperator;
+        this.scalingDownBlockedNodes = scalingDownBlockedNodes;
+    }
+
+    /**
+     * The main reconciliation method which triggers the whole reconciliation pipeline. This is the method which is
+     * expected to be called from the outside to trigger the reconciliation.
+     *
+     * @param kafkaStatus The Kafka Status class for updating auto-rebalancing status on it during the reconciliation
+     *
+     * @return  Future which completes when the reconciliation completes
+     */
+    public Future<Void> reconcile(KafkaStatus kafkaStatus) {
+        if (kafkaAutoRebalanceStatus != null) {
+            ScalingNodes scalingNodes = getScalingNodes(kafkaStatus.getAutoRebalance());
+            LOGGER.infoCr(reconciliation, "Reconciling auto-rebalance from state [{}] with scaling nodes: blocked = {}, added = {}",
+                    kafkaAutoRebalanceStatus.getState(), scalingNodes.blocked(), scalingNodes.added());
+            return computeNextStatus(kafkaAutoRebalanceStatus, scalingNodes)
+                    .onComplete(v -> kafkaStatus.setAutoRebalance(kafkaAutoRebalanceStatus));
+        } else {
+            LOGGER.infoCr(reconciliation, "No auto-rebalance state from the Kafka CR, initializing to [Idle]");
+            // when the auto-rebalance status doesn't exist, the Kafka cluster is being created
+            // so auto-rebalance can be set in an Idle state, no further actions
+            kafkaStatus.setAutoRebalance(
+                    new KafkaAutoRebalanceStatusBuilder()
+                            .withState(KafkaAutoRebalanceState.Idle)
+                            .withLastTransitionTime(StatusUtils.iso8601Now())
+                            .build());
+            return Future.succeededFuture();
+        }
```

Comment:
TBH, I'm still struggling to understand:
* Why do we need so much logic here. As I said, the other `*Reconciler` classes normally do preparation int he constructor and not in the `reconcile` method
* Given the above, I also struggle to see what the actual reconciliation is and what is it that does the work.

I think sticking with the structure of other classes named `*Reocnciler` would help the readability.

Reply:
AFAICS the other `*Reconciler* classes have a bunch of `compose` (because of async calls) to chain the actions for the reconciliation where it means for me dealing with what is in the "assembly" resource and check/create/update other corresponding resources (k8s or other custom ones).
In this case this role is played by the `computeNextStatus` (if you want we can rename it) which is in charge to call the right method based on the auto-rebalancing state and then dealing with the `KafkaRebalance` resource to update the status of the overall auto-rebalancing process. It actually sounds a reconciliation to me.

Regarding the `else` part of this method, this is the only way I have (or I figured out) to understand that it's the creation of the cluster and not a scale up/down (so the auto-rebalance status is still missing from the Kafka CR when it's cluster creation reconciliation) and the code added it here (which happens after the cluster creation). Maybe, I could try to pull this out in the `KafkaAssemblyOperator` and avoid to call the auto-rebalance reconciler at all if we are in the cluster creation stage.

Reply:
I think better name would be great. Because `computeNextStatus` absolutely doesn't suggest that it takes some action and not just computes the next status. Maybe you can call it something like `maybeRebalance` or something like that to suggest that this is where the action happens.

Reply:
Good idea. Renaming it.